src/service/storage-impl.js

```diff
@@ -0,0 +1,288 @@
+    this.binding_ = binding;
+
+    /** @const @private {string} */
+    this.origin_ = getSourceOrigin(this.win.location);
+
+    /** @const @private {boolean} */
+    this.isExperimentOn_ = isExperimentOn(this.win, EXPERIMENT);
+
+    /** @private @const {!Promise} */
+    this.whenStarted_ = this.isExperimentOn_ ?
+        Promise.resolve() :
+        Promise.reject(`Enable experiment ${EXPERIMENT}`);
+  }
+
+  /**
+   * @return {!Storage}
+   * @private
+   */
+  start_() {
+    if (!this.isExperimentOn_) {
+      log.info(TAG, 'Storage experiment is off: ', EXPERIMENT);
+      return this;
+    }
+    this.listenToBroadcasts_();
+    return this;
+  }
+
+  /**
+   * Returns the promise that yields the value of the property for the specified
+   * key.
+   * @param {string} name
+   * @return {!Promise<*>}
+   * @override
+   */
+  get(name) {
+    return this.getStore_().then(store => {
+      return store[name];
+    });
+  }
+
+  /**
+   * Saves the value of the specified property. Returns the promise that's
+   * resolved when the operation completes.
+   * @param {string} name
+   * @param {*} value
+   * @return {!Promise}
+   * @override
+   */
+  set(name, value) {
+    return this.saveStore_(store => {
+      store[name] = value;
+    });
+  }
+
+  /**
+   * Saves the specified property. Returns the promise that's resolved when
+   * the operation completes.
+   * @param {string} name
+   * @return {!Promise}
+   * @override
+   */
+  remove(name) {
+    return this.saveStore_(store => {
+      delete store[name];
+    });
+  }
+
+  /**
+   * @return {!Promise<!JSONObject>}
+   * @private
+   */
+  getStore_() {
+    if (this.store_) {
+      return Promise.resolve(this.store_);
+    }
+    return this.whenStarted_.then(() => {
+      return this.binding_.loadStore(this.origin_);
+    }).then(store => {
+      return this.store_ = store;
+    });
+  }
+
+  /**
+   * @param {function(!JSONObject)} mutator
+   * @return {!Promise}
+   * @private
+   */
+  saveStore_(mutator) {
+    return this.getStore_().then(store => {
+      mutator(store);
+      return this.binding_.saveStore(this.origin_, this.store_);
+    }).then(this.broadcastReset_.bind(this));
+  }
+
+  /** @private */
+  listenToBroadcasts_() {
+    this.viewer_.onBroadcast(message => {
+      if (message['type'] == 'amp-storage-reset' &&
+              message['origin'] == this.origin_) {
+        log.fine(TAG, 'Received reset message');
+        this.store_ = null;
+      }
+    });
+  }
+
+  /** @private */
+  broadcastReset_() {
+    log.fine(TAG, 'Broadcasted reset message');
+    this.viewer_.broadcast({
+      'type': 'amp-storage-reset',
+      'origin': this.origin_
+    });
+  }
+}
+
+
+/**
+ * A binding provides the specific implementation of storage technology.
+ * @interface
+ */
+class StorageBindingDef {
+
+  /**
+   * Returns the promise that yields the store blob for the specified origin.
+   * @param {string} unusedOrigin
```

Comment:
Yes, very unfortunately. The linter doesn't understand that this is an interface :(